design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+# External Authorization Support
+
+Status: Reviewing
+
+## Abstract
+This document describes a design for performing request authorization for virtual hosts hosted by Contour.
+
+## Background
+
+- [#432][5] Provide support for integrating with an external auth service.
+- [#2459][3] External auth background and goals.
+- [#2325][9] Failures in External services configured in envoy should be visible to the application developer.
+
+## Goals
+- Allow operators to integrate existing authorization services with Contour.
+- Allow arbitrary types of authentication and authorization to be supported.
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
+
+- Support for multiple external authorization mechanisms (i.e. the protocol the proxy uses to communicate with the authorization server).
+  Various proxies have different mechanisms for integrating external services, but it is impractical to support them all.
+
+- Abstracting the Envoy authorization mechanism.
+  The scope of abstracting the Envoy authorization mechanism is too large to be tractable.
+  Abstracting the protocol would also work against the goal of being able to integrate existing authorization servers.
+
+## High-Level Design
+A new `SupportService` CRD adds a way to represent and track an authorization service.
+This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.
+The core of the `SupportService` CRD is subset of the `projectcontour.v1.HTTPProxy` `Service` specification.
+Re-using the `Service` type allows the operator to specify configuration in familiar and consistent terms, especially TLS configuration.
+
+Note that only the Envoy [GRPC authorization protocol][2] will be supported.
+The GRPC protocol is a superset of the HTTP protocol and requires less configuration.
+The drawback of only supporting the GRPC protocol is that many existing Envoy authorization servers only support the HTTP protocol.
+Until Contour adds support for the v3 Envoy API, only the version 2 of the GRPC authorization protocol can be configured.
+
+Operators can bind an authorization service to a `HTTPProxy` using a new field in the `VirtualHost` struct.
```

Comment:
if I'm understanding correctly, this implies that all routes for a given vhost must use the same auth service (or disable it). Is that workable for users? Are there cases where users would want to use a different auth service for different subsets of the routes under a vhost?

Reply:
It's possible someone would like more flexibility than that, but I've yet to hear from that person. There's no way to specify this at a finer granularity other than funneling everything through an intermediate proxy that we configure.